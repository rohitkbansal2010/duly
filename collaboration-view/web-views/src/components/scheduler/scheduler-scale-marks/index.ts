export { SchedulerScaleMarks } from './scheduler-scale-marks';
export type { SchedulerScaleMarksPropsType } from './scheduler-scale-marks';
