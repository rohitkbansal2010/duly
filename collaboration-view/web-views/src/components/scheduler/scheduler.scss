@import "~@styles/colors.scss";
@import "~@styles/media-mixins.scss";
@import "~@styles/typography.scss";

.scheduler {
  position: relative;

  &Translucent {
    opacity: 0.2;
  }

  &GroupAppointments {
    position: absolute !important;
    left: 7.06rem;
    width: calc(100% - 7.06rem);
    margin: 0 !important;

    & > div {
      position: relative;
      margin: 0 !important;
      padding: 0 !important;

      &:not(:last-child) {
        margin-right: 0.31rem !important;
      }
    }
  }
}
