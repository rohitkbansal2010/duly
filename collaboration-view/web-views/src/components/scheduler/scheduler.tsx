import { useEffect, useState, useRef } from 'react';
import { Row } from 'react-bootstrap';
import { useSelector } from 'react-redux';

import { AppointmentItem } from '@components/appointment-item';
import { BTGModal } from '@components/btg-modal';
import { CurrentTimeBar } from '@components/current-time-bar';
import {
  HEIGHT_HOUR_SEGMENT,
  MINUTES_PER_HOUR,
  PADDING_BOTTOM_SCHEDULER_SCALE
} from '@constants';
import { RootState } from '@redux/reducers';
import { GroupedAppointmentType } from '@types';
import {
  calcDurationInMinutes,
  calcTopOfGroupsAppointems,
  cropAppointmentsByWorkDay,
  groupAppointments,
  sortAppointmentsInDescTime
} from '@utils';

import { PickedAppointmentModal } from '../picked-appointment-modal';

import { SchedulerScaleMarks } from './scheduler-scale-marks';
import styles from './scheduler.scss';

const workDayStart = window?.env?.WORK_DAY_START || '';
const workDayEnd = window?.env?.WORK_DAY_END || '';

export const Scheduler = () => {
  const appointments = useSelector(({ TODAYS_APPOINTMENTS }: RootState) =>
    TODAYS_APPOINTMENTS.appointments);
  const pickedAppointmentId = useSelector(({ TODAYS_APPOINTMENTS }: RootState) =>
    TODAYS_APPOINTMENTS.pickedAppointmentId);
  const isSpinnerShown = useSelector(({ APPSTATE }: RootState) =>
    APPSTATE.isSpinnerShown);

  const schedulerClass = []
    .concat(styles.scheduler)
    .concat(isSpinnerShown ? styles.schedulerTranslucent : '')
    .join(' ');

  const [ groupsAppointments, setGrouspAppointments ] = useState<GroupedAppointmentType[]>([]);

  const schedulerRef = useRef<HTMLDivElement>(null);

  const scrollToScheduler = () => {
    schedulerRef.current?.scrollIntoView({
      behavior: 'smooth',
      block: 'center',
    });
  };

  useEffect(() => {
    scrollToScheduler();
  }, [ groupsAppointments ]);

  const renderGroupsAppointments =
    ({ startTime, endTime, appointments }: GroupedAppointmentType, i: number) => {
      const groupsDuration = calcDurationInMinutes(new Date(startTime), new Date(endTime));
      const groupsHeight =
        (groupsDuration / MINUTES_PER_HOUR) * HEIGHT_HOUR_SEGMENT - PADDING_BOTTOM_SCHEDULER_SCALE;

      return (
        <Row
          ref={i === 0 ? schedulerRef : undefined}
          key={startTime}
          className={styles.schedulerGroupAppointments}
          style={{
            top: `${calcTopOfGroupsAppointems(new Date(startTime))}rem`,
            height: `${groupsHeight}rem`,
          }}
        >
          {appointments.map(({ id, title, ...props }) =>
            (
              <AppointmentItem
                key={id}
                id={id}
                title={title}
                startTimeGroups={startTime}
                {...props}
              />
            ))}
        </Row>
      );
    };

  useEffect(() => {
    const groupedAppointments = groupAppointments(cropAppointmentsByWorkDay(appointments))
      .map(groupedAppointment =>
        ({
          ...groupedAppointment,
          appointments: sortAppointmentsInDescTime(groupedAppointment.appointments),
        }));

    setGrouspAppointments(groupedAppointments);
  }, [ appointments ]);

  return (
    <div className={schedulerClass}>
      {pickedAppointmentId && (
        <PickedAppointmentModal />
      )}
      <BTGModal />
      <SchedulerScaleMarks
        workDayStart={+workDayStart}
        workDayEnd={+workDayEnd}
      />
      <CurrentTimeBar />
      {groupsAppointments.map(renderGroupsAppointments)}
    </div>
  );
};
