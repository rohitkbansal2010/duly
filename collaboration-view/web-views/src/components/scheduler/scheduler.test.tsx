import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as redux from 'react-redux';

import { AppointmentStatus, AppointmentType, PractitionerRole } from '@enums';
import { AppointmentData } from '@types';

import { Scheduler } from './scheduler';

const mockAppointments: AppointmentData[] = [
  {
    id: 'qwerty2',
    title: 'Check Up',
    type: AppointmentType.TELEHEALTH,
    status: AppointmentStatus.ARRIVED,
    timeSlot: {
      startTime: '2021-12-23T11:00:00',
      endTime: '2021-12-23T11:15:00',
    },
    location: { title: 'Room #3' },
    patient: {
      patientGeneralInfo: {
        id: 'qwerty2',
        humanName: {
          familyName: 'Smith',
          givenNames: [ 'Benjamin' ],
        },
      },
      isNewPatient: false,
    },
    practitioner: {
      id: 'qwerty2',
      humanName: {
        familyName: 'Sussman',
        prefixes: [ 'Dr.' ],
      },
      role: PractitionerRole.PCP,
    },
  },
  {
    id: 'qwerty3',
    title: 'Check Up',
    type: AppointmentType.ON_SITE,
    status: AppointmentStatus.IN_PROGRESS,
    timeSlot: {
      startTime: '2021-12-23T11:00:00',
      endTime: '2021-12-23T11:30:00',
    },
    location: { title: 'Room #2' },
    patient: {
      patientGeneralInfo: {
        id: 'qwerty3',
        humanName: {
          familyName: 'Johnson',
          givenNames: [ 'Mary' ],
        },
      },
      isNewPatient: true,
    },
    practitioner: {
      id: 'qwerty3',
      humanName: { familyName: 'Palmero' },
      role: PractitionerRole.MA,
    },
  },
];

describe('Scheduler', () => {
  let component: ReturnType<typeof shallow>;

  beforeEach(() => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((selector: (state: unknown) => unknown) =>
        selector({
          TODAYS_APPOINTMENTS: { appointments: mockAppointments },
          APPSTATE: {},
        }));
    jest
      .spyOn(redux, 'useDispatch')
      .mockReturnValue(jest.fn());

    component = shallow(<Scheduler />);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should exist', () => {
    expect(Scheduler).toBeDefined();
  });

  it('should render the components SchedulerScaleMarks, CurrentTimeBar', () => {
    expect(toJson(component)).toMatchSnapshot();

    expect(component.find('SchedulerScaleMarks')).toHaveLength(1);
    expect(component.find('CurrentTimeBar')).toHaveLength(1);
  });
});
