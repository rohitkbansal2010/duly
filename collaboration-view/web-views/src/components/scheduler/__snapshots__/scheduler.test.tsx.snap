// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduler should render the components SchedulerScaleMarks, CurrentTimeBar 1`] = `
<div
  className="scheduler "
>
  <BTGModal />
  <SchedulerScaleMarks
    workDayEnd={19}
    workDayStart={7}
  />
  <CurrentTimeBar />
</div>
`;
